docs/source/miscellaneous/doc/adr/0017-transactions3.md

```diff
@@ -0,0 +1,302 @@
+followed by an extended failure or network partition could result in inconsistent reads. In particular, supposing a 
+three-node cluster `[A, B, C]`, on a successful CAS of a value from `V1` to `V2` where A was the coordinator and only 
+successfully applied the mutation to itself, a quorum read of this cell could read `V1` (if the quorum chosen was B, C) 
+*or* `V2` (if the quorum included A: this value is newer, assuming no weird write-time shenanigans).
+
+This is a problem, because in AtlasDB we rely on putUnlessExists for the transactions table (which uses Cassandra's 
+CAS). In practice, this means that a transaction at a given timestamp could be read both as being committed and being 
+uncommitted. Furthermore, it is possible that a quorum of nodes that believe the transaction is uncommitted decide that 
+the transaction should be rolled back (as they would accept a putUnlessExists of `-1` to that cell, which is interpreted
+in Atlas as meaning the transaction was aborted).
+
+### Range Movements
+
+Cassandra determines which nodes participate in the Paxos based on the replication group of the relevant key. This, 
+however, may change at runtime (e.g., if there are changes in cluster topology), which can naturally lead to disjoint 
+quorums. Cassandra is not currently resistant to this case, and this would require substantial changes to the Paxos 
+algorithm. We're thus treating dealing with these cases as outside the scope of this workstream.
+
+## Decision
+
+### Criteria
+
+A good solution to this problem should demonstrate the following characteristics:
+
+- *Repeatable reads*: After a non-null value (a commit timestamp, or ABORTED) has been read from the table for a given 
+  timestamp, all future reads should return the same value. This is our primary issue with the current system.
+- *Eventual read efficiency*: Eventually, reading a cell in the transactions table should require just 1 RPC from
+  AtlasDB to the coordinating Cassandra node, and within Cassandra just 1 RPC between each pair of Cassandra nodes in
+  the relevant replication group.
+- *Write efficiency*: Writing a cell in the transactions table should require as few RPCs as possible, bearing in mind 
+  the other constraints. Optimising for this property is less important than the other two.
+
+### Theory
+
+#### PUE Tables
+We want to define a *put-unless-exists table*, which supports only two operations: putUnlessExists and get. While this
+interface may look superfluous in that the `KeyValueService` interface already has very similar endpoints, it *must*
+support repeatable reads.
+
+```java
+public interface PutUnlessExistsTable<K, V> {
+    void putUnlessExistsMultiple(Map<K, V> keyValues) throws KeyAlreadyExistsException;
+
+    ListenableFuture<Map<K, V>> get(Iterable<K> keys);
+}
+```
+
+As a primitive, we define a *put-unless-exists value* as a pair of a value and a state enumeration. This state indicates
+whether the value is *pending* or *committed*.
```

Comment:
To match the below:
```suggestion
whether the value is *staging* or *committed*.
```

(I actually prefer 'pending' throughout, but not strongly)

Reply:
Ah. Yeah it has to be _STAGING_, good spot